File: organism/genetics.py
from __future__ import annotations
from enum import Enum, auto
from random import choice, uniform
from organism.ontology import Temperament, Diet
from decisions.actions import  MoveActions
from core.coord import Coord
from dataclasses import dataclass
from abc import ABC, abstractmethod


class CreatureTypes(Enum):
    CRAB = auto()
    FISH = auto()
    CROCODILE = auto()
    HIPPOPOTAMUS = auto()

    @staticmethod
    def choice() -> CreatureTypes:
        return choice([CreatureTypes.CRAB, CreatureTypes.FISH, CreatureTypes.CROCODILE, CreatureTypes.HIPPOPOTAMUS])
    



def gen_disturb(v_min:float, v_max:float) -> float:
    return uniform(v_min, v_max)
def gen_little_disturb() -> float:
    return gen_disturb(0.90, 1.10)
def gen_medium_disturb() -> float:
    return gen_disturb(0.86, 1.16)
def smooth_scramble(v1:int | float, v2:int | float) -> int | float:
    return (v1 + v2)/2 * gen_little_disturb()
def time_scramble(v1:int, v2:int) -> int:
    return round((v1 + v2)/2) + choice([-1, 0, 1])




class BaseGenome(ABC):
    
    @abstractmethod
    def scramble(self, other) -> BaseGenome:...


@dataclass(frozen=True)
class MetabolismGenome(BaseGenome):
    max_hungry:float
    diet:Diet
    mass:int | float
    energy_limit:int | float

    def scramble(self, other:MetabolismGenome) -> MetabolismGenome:
        return MetabolismGenome(
            smooth_scramble(self.max_hungry, other.max_hungry),
            self.diet.scramble(other.diet),
            self.mass * gen_little_disturb(),
            (self.energy_limit + other.energy_limit)/2 * gen_medium_disturb()
        )

@dataclass(frozen=True)
class ReproductionGenome(BaseGenome):
    reproduction_cost:int | float
    extra_reproduction_multiplier: int | float
    fertility_limit:int
    gestation_time:int
    
    def scramble(self, other:ReproductionGenome) -> ReproductionGenome:
        return ReproductionGenome(
            smooth_scramble(self.reproduction_cost, other.reproduction_cost),
            smooth_scramble(self.extra_reproduction_multiplier, other.extra_reproduction_multiplier),
            time_scramble(self.fertility_limit, other.fertility_limit),
            time_scramble(self.gestation_time, other.gestation_time)
        )
    
@dataclass(frozen=True)
class BodyGenome(BaseGenome):
    life_limit:int | float
    strength: int | float
    max_age: int

    def scramble(self, other:BodyGenome) -> BodyGenome:
        return BodyGenome(
            smooth_scramble(self.life_limit, other.life_limit),
            smooth_scramble(self.strength, other.strength),
            time_scramble(self.max_age, other.max_age)
        )

@dataclass(frozen=True)
class CoreGenome:
    capabilities:set[MoveActions]
    vision_radius:Coord
    behavior:Temperament
    id:CreatureTypes
  

@dataclass(frozen=True)
class Genome:
    metabolism:MetabolismGenome
    body:BodyGenome
    reproduction:ReproductionGenome
    core:CoreGenome

    def crossover(self, other:Genome) -> Genome:
        new_metabolism_g = self.metabolism.scramble(other.metabolism)
        new_body_g = self.body.scramble(other.body)
        new_reproduction_g = self.reproduction.scramble(other.reproduction)
        
        return Genome(
            new_metabolism_g,
            new_body_g,
            new_reproduction_g,
            self.core
        )

class CreatureGenomes:
    g_metabolism: dict[CreatureTypes, MetabolismGenome] = {}
    g_body: dict[CreatureTypes, BodyGenome] = {}
    g_reproduction: dict[CreatureTypes, ReproductionGenome] = {}
    g_core: dict[CreatureTypes, CoreGenome] = {}

    def exists_creature_type(self, c_t: CreatureTypes) -> bool:
        return (
            c_t in self.g_metabolism
            and c_t in self.g_body
            and c_t in self.g_reproduction
            and c_t in self.g_core
        )

    def get_type_genome(self, c_t: CreatureTypes) -> Genome:
        if not self.exists_creature_type(c_t):
            raise ValueError(f'Creature type {c_t} does not exist in all genomes')
        return Genome(
            self.g_metabolism[c_t],
            self.g_body[c_t],
            self.g_reproduction[c_t],
            self.g_core[c_t]
        )

    def add_g_metabolism(self, g_metabolism: MetabolismGenome, c_t: CreatureTypes) -> None:
        self.g_metabolism[c_t] = g_metabolism

    def add_g_body(self, g_body: BodyGenome, c_t: CreatureTypes) -> None:
        self.g_body[c_t] = g_body

    def add_g_reproduction(self, g_reproduction: ReproductionGenome, c_t: CreatureTypes) -> None:
        self.g_reproduction[c_t] = g_reproduction

    def add_g_core(self, g_core: CoreGenome, c_t: CreatureTypes) -> None:
        self.g_core[c_t] = g_core
    

def populate_default_genomes(creatures_genomes: CreatureGenomes) -> None:
    # CROCODILE
    creatures_genomes.add_g_metabolism(MetabolismGenome(
        max_hungry=0.80,
        diet=Diet(0.58, 0.085, 0.85),
        mass=1.25,
        energy_limit=100
    ), CreatureTypes.CROCODILE)
    creatures_genomes.add_g_body(BodyGenome(
        life_limit=100,
        strength=27,
        max_age=50
    ), CreatureTypes.CROCODILE)
    creatures_genomes.add_g_reproduction(ReproductionGenome(
        reproduction_cost=40,
        extra_reproduction_multiplier=1.2,
        fertility_limit=5,
        gestation_time=5
    ), CreatureTypes.CROCODILE)
    creatures_genomes.add_g_core(CoreGenome(
        capabilities={MoveActions.WALK, MoveActions.SWIMM},
        vision_radius=Coord(4, 4),
        behavior=Temperament.AGGRESSIVE,
        id=CreatureTypes.CROCODILE
    ), CreatureTypes.CROCODILE)

    # CRAB
    creatures_genomes.add_g_metabolism(MetabolismGenome(
        max_hungry=0.65,
        diet=Diet(0.5, 0.35, 0.15),
        mass=1.1,
        energy_limit=40
    ), CreatureTypes.CRAB)
    creatures_genomes.add_g_body(BodyGenome(
        life_limit=30,
        strength=7.5,
        max_age=30
    ), CreatureTypes.CRAB)
    creatures_genomes.add_g_reproduction(ReproductionGenome(
        reproduction_cost=5,
        extra_reproduction_multiplier=1.2,
        fertility_limit=3,
        gestation_time=4
    ), CreatureTypes.CRAB)
    creatures_genomes.add_g_core(CoreGenome(
        capabilities={MoveActions.WALK},
        vision_radius=Coord(3, 3),
        behavior=Temperament.NEUTRAL,
        id=CreatureTypes.CRAB
    ), CreatureTypes.CRAB)

    # FISH
    creatures_genomes.add_g_metabolism(MetabolismGenome(
        max_hungry=0.33,
        diet=Diet(0.2, 0.75, 0.15),
        mass=1.0,
        energy_limit=10
    ), CreatureTypes.FISH)
    creatures_genomes.add_g_body(BodyGenome(
        life_limit=9,
        strength=1,
        max_age=10
    ), CreatureTypes.FISH)
    creatures_genomes.add_g_reproduction(ReproductionGenome(
        reproduction_cost=1.9,
        extra_reproduction_multiplier=1.0,
        fertility_limit=1,
        gestation_time=2
    ), CreatureTypes.FISH)
    creatures_genomes.add_g_core(CoreGenome(
        capabilities={MoveActions.SWIMM},
        vision_radius=Coord(2, 4),
        behavior=Temperament.PASSIVE,
        id=CreatureTypes.FISH
    ), CreatureTypes.FISH)

    # HIPPOPOTAMUS
    creatures_genomes.add_g_metabolism(MetabolismGenome(
        max_hungry=0.70,
        diet=Diet(0.2, 0.7, 0.1),
        mass=1.6,
        energy_limit=85
    ), CreatureTypes.HIPPOPOTAMUS)
    creatures_genomes.add_g_body(BodyGenome(
        life_limit=150,
        strength=20,
        max_age=60
    ), CreatureTypes.HIPPOPOTAMUS)
    creatures_genomes.add_g_reproduction(ReproductionGenome(
        reproduction_cost=40,
        extra_reproduction_multiplier=1.3,
        fertility_limit=4,
        gestation_time=5
    ), CreatureTypes.HIPPOPOTAMUS)
    creatures_genomes.add_g_core(CoreGenome(
        capabilities={MoveActions.WALK, MoveActions.SWIMM},
        vision_radius=Coord(3, 3),
        behavior=Temperament.TERRITORIAL,
        id=CreatureTypes.HIPPOPOTAMUS
    ), CreatureTypes.HIPPOPOTAMUS)


creatures_genomes = CreatureGenomes()
populate_default_genomes(creatures_genomes)

File: src/karkinolution/organism/genetics.cpp
#include <karkinolution/organism/genetics.hpp>
#include <karkinolution/utils/k_random.hpp>

// CreatureSpeciesF

CreatureSpecies CreatureSpeciesF::choice() {
    return Choices::choice({
        CreatureSpecies::CRAB, 
        CreatureSpecies::CROCODILE,
        CreatureSpecies::FISH, 
        CreatureSpecies::HIPPOPOTAMUS
    });
}

// GenomeMotor
MetabolismGenome GenomeMotor::scramble(const MetabolismGenome& g1, const MetabolismGenome& g2) {
    return MetabolismGenome{
        .max_hungry = Disturbs::smooth_scramble(g1.max_hungry, g2.max_hungry),
        .mass = Disturbs::gen_little_disturb() * g1.mass,
        .energy_limit = (g1.energy_limit + g2.energy_limit) / 2 * Disturbs::gen_medium_disturb(),
        .diet = g1.diet.scramble(g2.diet)
    };
}

ReproductionGenome GenomeMotor::scramble(const ReproductionGenome& g1, const ReproductionGenome& g2) {
    return ReproductionGenome{
        .reproduction_cost = Disturbs::smooth_scramble(g1.reproduction_cost, g2.reproduction_cost),
        .extra_reproduction_multiplier = Disturbs::smooth_scramble(g1.extra_reproduction_multiplier, g2.extra_reproduction_multiplier),
        .fertility_limit = Disturbs::time_scramble(g1.fertility_limit, g2.fertility_limit),
        .gestation_time_limit = Disturbs::time_scramble(g1.gestation_time_limit, g2.gestation_time_limit)
    };
}

BodyGenome GenomeMotor::scramble(const BodyGenome& g1, const BodyGenome& g2) {
    return BodyGenome{
        .life_limit = Disturbs::smooth_scramble(g1.life_limit, g2.life_limit),
        .strength = Disturbs::smooth_scramble(g1.strength, g2.strength),
        .max_age = Disturbs::time_scramble(g1.max_age, g2.max_age)
    };
}

Genome GenomeMotor::crossover(const Genome& g1, const Genome& g2) {
    return Genome{
        .metabolism = scramble(g1.metabolism, g2.metabolism),
        .reproduction = scramble(g1.reproduction, g2.reproduction),
        .body = scramble(g1.body, g2.body),
        .core = g1.core
    };
}

// CreatureGenoms

bool CreatureGenomes::exists_specie(CreatureSpecies c_s) const {
    return g_metabolism.contains(c_s) &&
           g_body.contains(c_s) &&
           g_reproduction.contains(c_s) &&
           g_core.contains(c_s);
}

Genome CreatureGenomes::get_genome(CreatureSpecies c_s) const {
    return Genome{
        .metabolism = g_metabolism.at(c_s),
        .reproduction = g_reproduction.at(c_s),
        .body = g_body.at(c_s),
        .core = g_core.at(c_s)
    };
}

void CreatureGenomes::add(CreatureSpecies c_s, const MetabolismGenome& g) {
    g_metabolism.insert_or_assign(c_s, g);
}

void CreatureGenomes::add(CreatureSpecies c_s, const BodyGenome& g) {
    g_body.insert_or_assign(c_s, g);
}

void CreatureGenomes::add(CreatureSpecies c_s, const ReproductionGenome& g) {
    g_reproduction.insert_or_assign(c_s, g);
}

void CreatureGenomes::add(CreatureSpecies c_s, const CoreGenome& g) {
    g_core.insert_or_assign(c_s, g);
}
// Populate

// ==========================================
// Populate Namespace
// ==========================================

void Populate::populate_crocodile(CreatureGenomes& creature_genomes) {
    creature_genomes.add(CreatureSpecies::CROCODILE, MetabolismGenome{
        .max_hungry = 0.80f,
        .mass = 1.25f,
        .energy_limit = 100.0f,
        .diet = Diet(0.58f, 0.085f, 0.85f) // Assumindo construtor Diet(float, float, float)
    });

    creature_genomes.add(CreatureSpecies::CROCODILE, BodyGenome{
        .life_limit = 100.0f,
        .strength = 27.0f,
        .max_age = 50
    });

    creature_genomes.add(CreatureSpecies::CROCODILE, ReproductionGenome{
        .reproduction_cost = 40.0f,
        .extra_reproduction_multiplier = 1.2f,
        .fertility_limit = 5,
        .gestation_time_limit = 5
    });

    creature_genomes.add(CreatureSpecies::CROCODILE, CoreGenome{
        .capabilities = {MoveActions::WALK, MoveActions::SWIMM},
        .vision_radius = Vec2(4, 4),
        .temperament = Temperament::AGGRESSIVE,
        .specie = CreatureSpecies::CROCODILE
    });
}

void Populate::populate_fish(CreatureGenomes& creature_genomes) {
    creature_genomes.add(CreatureSpecies::FISH, MetabolismGenome{
        .max_hungry = 0.33f,
        .mass = 1.0f,
        .energy_limit = 10.0f,
        .diet = Diet(0.2f, 0.75f, 0.15f)
    });

    creature_genomes.add(CreatureSpecies::FISH, BodyGenome{
        .life_limit = 9.0f,
        .strength = 1.0f,
        .max_age = 10
    });

    creature_genomes.add(CreatureSpecies::FISH, ReproductionGenome{
        .reproduction_cost = 1.9f,
        .extra_reproduction_multiplier = 1.0f,
        .fertility_limit = 1,
        .gestation_time_limit = 2
    });

    creature_genomes.add(CreatureSpecies::FISH, CoreGenome{
        .capabilities = {MoveActions::SWIMM},
        .vision_radius = Vec2(2, 4),
        .temperament = Temperament::PASSIVE,
        .specie = CreatureSpecies::FISH
    });
}

void Populate::populate_crab(CreatureGenomes& creature_genomes) {
    creature_genomes.add(CreatureSpecies::CRAB, MetabolismGenome{
        .max_hungry = 0.65f,
        .mass = 1.1f,
        .energy_limit = 40.0f,
        .diet = Diet(0.5f, 0.35f, 0.15f)
    });

    creature_genomes.add(CreatureSpecies::CRAB, BodyGenome{
        .life_limit = 30.0f,
        .strength = 7.5f,
        .max_age = 30
    });

    creature_genomes.add(CreatureSpecies::CRAB, ReproductionGenome{
        .reproduction_cost = 5.0f,
        .extra_reproduction_multiplier = 1.2f,
        .fertility_limit = 3,
        .gestation_time_limit = 4
    });

    creature_genomes.add(CreatureSpecies::CRAB, CoreGenome{
        .capabilities = {MoveActions::WALK},
        .vision_radius = Vec2(3, 3),
        .temperament = Temperament::NEUTRAL,
        .specie = CreatureSpecies::CRAB
    });
}

void Populate::populate_hippopotamus(CreatureGenomes& creature_genomes) {
    creature_genomes.add(CreatureSpecies::HIPPOPOTAMUS, MetabolismGenome{
        .max_hungry = 0.70f,
        .mass = 1.6f,
        .energy_limit = 85.0f,
        .diet = Diet(0.2f, 0.7f, 0.1f)
    });

    creature_genomes.add(CreatureSpecies::HIPPOPOTAMUS, BodyGenome{
        .life_limit = 150.0f,
        .strength = 20.0f,
        .max_age = 60
    });

    creature_genomes.add(CreatureSpecies::HIPPOPOTAMUS, ReproductionGenome{
        .reproduction_cost = 40.0f,
        .extra_reproduction_multiplier = 1.3f,
        .fertility_limit = 4,
        .gestation_time_limit = 5
    });

    creature_genomes.add(CreatureSpecies::HIPPOPOTAMUS, CoreGenome{
        .capabilities = {MoveActions::WALK, MoveActions::SWIMM},
        .vision_radius = Vec2(3, 3),
        .temperament = Temperament::TERRITORIAL,
        .specie = CreatureSpecies::HIPPOPOTAMUS
    });
}

